src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs

```diff
@@ -1125,4 +1127,146 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
             codeStream.Emit(ILOpcode.call, emitter.NewToken(helper));
         }
     }
+
+    class CustomTypeMarshaller : Marshaller
+    {
+        protected override void TransformManagedToNative(ILCodeStream codeStream)
+        {
+            var marshallerType = MarshalAsDescriptor.CustomMarshallerType;
+            if (marshallerType == null)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (marshallerType.IsGenericDefinition)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowTypeLoadException(marshallerType);
+            }
+
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            if (customMarshallerType == null)
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (!marshallerType.CanCastTo(customMarshallerType))
+            {
+                // ApplicationException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            var getInstanceMethod = marshallerType?.GetMethod(
```

Comment:
Can `marshallerType` be null here? We are using the variable number of times above.

Reply:
I working on this right now. My idea is to pass Type to InitializeCustomMarshaller and let it decide what exceptions to throw. That allow me to cover all cases except one.